<a href="https://colab.research.google.com/github/yywu123/CSC44800----Artificial-Intelligence/blob/main/Assignment%203%20%3A%20Equation%20of%20a%20Slime%20/Assignment3%20.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


 # <center> Assignment 3 :  Equation of a Slime  </center>

## Import Liberay 

In [159]:
# Imports section
import pandas as pd 
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedKFold




## Part 1. Loading the dataset


In [4]:
# Using pandas load the dataset (load remotely, not locally)
dataset = pd.read_csv("https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv")

In [6]:
# Output the first 15 rows of the data
dataset.head(15)

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [8]:
# Display a summary of the table information (number of datapoints, etc.)
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


In [135]:
dataset.describe()

Temperature °C     Mols KCL     Size nm^3
count     1000.000000  1000.000000  1.000000e+03
mean       500.500000   471.530000  5.086111e+05
std        288.819436   288.482872  4.474838e+05
min          1.000000     1.000000  1.611429e+01
25%        250.750000   226.750000  1.298267e+05
50%        500.500000   459.500000  3.827182e+05
75%        750.250000   710.250000  7.603211e+05
max       1000.000000  1000.000000  1.972127e+06

---

 ## Part 2. Splitting the dataset

In [78]:
# Take the pandas dataset and split it into our features (X) and label (y)


x = dataset[['Temperature °C','Mols KCL']]
y = dataset[['Size nm^3']]

columns = dataset.columns
print("Columns Name: ")
columns.tolist()

Columns Name: 


['Temperature °C', 'Mols KCL', 'Size nm^3']

According the summary in the part 1, there are three colums : Temperature °C , Mols KCL and Size nm^3. The temperature and mole are independent variable, so I put these two together in x. The size of Slime is the dependent variable, so I put this in y. 

In [42]:
# Output the first 15 rows of the x 
x.head(15)

Temperature °C  Mols KCL
0              469       647
1              403       694
2              302       975
3              779       916
4              901        18
5              545       637
6              660       519
7              143       869
8               89       461
9              294       776
10             991       117
11             307       781
12             206        70
13             437       599
14             566        75

In [43]:
# Output the first 15 rows of the y
y.head(15)

Size nm^3
0   6.244743e+05
1   5.779610e+05
2   6.196847e+05
3   1.460449e+06
4   4.325726e+04
5   7.124634e+05
6   7.006960e+05
7   2.718260e+05
8   8.919803e+04
9   4.770210e+05
10  2.441771e+05
11  5.006455e+05
12  3.145200e+04
13  5.390215e+05
14  9.185271e+04

In [71]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.1, train_size = 0.9, random_state=100)

X_train

Temperature °C  Mols KCL
630             651       147
904             107       140
239             157       145
508             301       885
833               2         3
..              ...       ...
855             510       256
871              14        43
835               6        13
792             175       236
520             456       188

[900 rows x 2 columns]

In [70]:
y_train

Size nm^3
630  199823.400000
904   31804.000000
239   48014.714290
508  558759.857100
833      36.257143
..             ...
855  269112.457100
871    1424.828571
835     232.828571
792   86291.314290
520  177937.828600

[900 rows x 1 columns]

In [69]:
X_test

Temperature °C  Mols KCL
249             924       140
353             466       127
537             159       296
424             321        58
564              64       311
..              ...       ...
559             726       744
134             920       615
101             186       829
265             300       555
301             306       169

[100 rows x 2 columns]

In [68]:
y_test

Size nm^3
249  2.703680e+05
353  1.244168e+05
537  9.853931e+04
424  4.118411e+04
564  4.333946e+04
..            ...
559  1.104815e+06
134  1.153446e+06
101  3.302555e+05
265  3.454007e+05
301  1.079160e+05

[100 rows x 1 columns]

---
## Part 3. Perform a Linear Regression

In [75]:
# Use sklearn to train a model on the training set

reg = LinearRegression().fit(X_train, y_train)



In [114]:
# Create a sample datapoint and predict the output of that sample with the trained model

## A sample datapoint that I randonly select from the test dataset  
print("Simple datapoint information: ")
sample_datapoint = X_test[X_test['Temperature °C'] == 726]
print(dataset[dataset['Temperature °C'] == 726])

print("Prrediction using trained moedel")
prediction = reg.predict(sample_datapoint)
print(prediction)

Simple datapoint information: 
     Temperature °C  Mols KCL    Size nm^3
559             726       744  1104815.314
Prrediction using trained moedel
[[988422.26119434]]


In [147]:
# Report on the score for that model, in your own words (markdown, not code) explain what the score means

train_score = reg.score(X_train,y_train)
test_score = reg.score(X_test, y_test)
print(train_score)
print(test_score)


0.8601023011329709
0.8649421479766051


 As we know, the score() function from sklearn will return the coefficient of the determination. The best possible you can get from the coeffiecient of the determination is 1, which means the model can pefectly predict the dependent value. Therefore, if the score is really close to 1, it means your models is very good and accurate at predicting the result. 

The train_score is 0.86 and the test_score is 0.865. This means that the accurateness of our model is very high. Also, it indicate a strong positive linear relationship between temperature and moles with the size of the slime. 

In [156]:
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
# Note: When writing your output equations for your sample outputs you can ignore outside of 5 significant figures.

coefficent = reg.coef_
intercept = reg.intercept_


print("Coefficent: ", coefficent)
print("Intercept: ", intercept)

Coefficent:  [[ 883.23686227 1034.42461006]]
Intercept:  [-422419.61069843]


### Write an equation for your h(x) using LaTex :
 When writing your output equations for your sample outputs you can ignore outside of 5 significant figures.
 
####  <center> $ h(x) = 883.23686 x_{1} + 1034.42451 x_{2} - 422419.61070 $ </center>

$x_{1}$ represent temperature and $x_{2}$ represent mols

---
## Part 4. Use Cross Validation

In [171]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data

## I will use repeat k fold 
rkf = RepeatedKFold(n_splits=10, n_repeats=100, random_state=1000)
cross_valid_scores = cross_val_score(reg,x,y,cv= rkf)
print(cross_valid_scores[:10])



[0.89299883 0.86398905 0.82648862 0.8225824  0.84794573 0.84951624
 0.86407048 0.87194134 0.86661917 0.87335083]


In [172]:
# Report on their finding and their significance

print("%0.2f accuracy with a standard deviation of %0.2f" % (cross_valid_scores.mean(), cross_valid_scores.std()))


0.86 accuracy with a standard deviation of 0.02


According to scikit learn, cross-validation is a method to evaluate the estimator's performance. Corss-validation will split the training set into k smaller set. The model is trained using k - 1 of the fold as training data. The resulting model is validated on the remaining part of the data. This process is call k-fold cross validation. 

Cross validation is important because it can avoid overfitting situation. Sometime, our model will do good at the label that repeat many time, but will do bad at the unseem label. We can use cross validation to check if our model have overfitting problem or not.

The result of cross validation with repeated k fold is 0.86 accuaracy with a standard deviation of 0.02. This can show that our model do not have overfitting problem because the train_score is 0.86 and the test_score is 0.865. It basically the same result. Also, the result of the cross validation is really close to 1, this also can prove our model is good. 

---
## Part 5. Using Polynomial Regression

In [152]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2

poly = PolynomialFeatures(degree = 2)
poly_transform_x = poly.fit_transform(x)

## split the data into a training/test set. (90% train, 10% test)
x_poly_train, x_poly_test, y_poly_train, y_poly_test = train_test_split(poly_transform_x, y, test_size=0.1, train_size = 0.9, random_state=100)

##  Use sklearn to train a model on the training set
reg_poly = LinearRegression().fit(x_poly_train, y_poly_train)


In [153]:
# Report on the metrics and output the resultant equation as you did in Part 3.

# score
train_score_poly = reg_poly.score(x_poly_train,y_poly_train)
test_score_poly = reg_poly.score(x_poly_test,y_poly_test)
print("Train score: ", train_score_poly)
print("Test score: ", test_score_poly)

Train score:  1.0
Test score:  1.0


The score for this model is 1. This means the model can pefectly predict the dependent value

In [155]:
poly_coefficent = reg_poly.coef_
poly_intercept = reg_poly.intercept_


print("Coefficent: ", poly_coefficent)
print("Intercept: ", poly_intercept)

Coefficent:  [[ 0.00000000e+00  1.20000000e+01 -1.41976606e-07 -4.50779144e-12
   2.00000000e+00  2.85714287e-02]]
Intercept:  [2.07960256e-05]


### Write an equation for your h(x) using LaTex :
#### [$x_{1}$,$ x_{2} $ ] to [1, $x_{1}, x_{2}, x_{1}^2, x_{1}x_{2}, x_{2}^{2} $ ]
When writing your output equations for your sample outputs you can ignore outside of 5 significant figures.

####  <center> $ h(x) = 0 + 1.2 * 10^1 x_{1} - 1.41977 * 10^{-7} x_{2} - 4.50779 10^{-12} x_{1}^2 + 2* x_{1}x_{2} + 2.85714 * 10^{-2} x_{2}^{2} + 2.0796 * 10^{-5}   $ </center>

#### simplify : 
####  <center> $ h(x) = 12x_{1} + 2x_{1}x_{2} + 0.028571x_{2}^{2} + 0.00002   $ </center>


    